ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java

```diff
@@ -114,7 +115,8 @@ public void installSnapshot(InstallSnapshotRequestProto request, StateMachine st
 
     // TODO: Make sure that subsequent requests for the same installSnapshot are coming in order,
     // and are not lost when whole request cycle is done. Check requestId and requestIndex here
-
+    MessageDigest digester = newMd5Digest();
+    digester.reset();
```

Comment:
Let's use `MD5Hash.getDigester()`.

Reply:
Since I am not sure if it is a problem with JavaUtils.memoize, but I can confirm that it is a problem with digest, my idea is to use a new digest each time, which is more secure and reliable, and also avoids the situation where the digest calculation buffer is destroyed and the final result is incorrect

Reply:
`MD5Hash.getDigester()` is safe since it is a `Threadlocal` and reset the `digest` every time, i.e. the same thread will always use the same `digest` object.

Reply:
Yes, ThreadLocal variables should be safe. However, we have stably reproduced this issue in multiple different testing environments, and we are also identifying whether there is a problem with the JDK version we are using. During this process, we have made an important discovery that even if we receive an empty file, Corrupt will still appear.

<img width="805" alt="image" src="https://github.com/user-attachments/assets/7cacaa35-68cd-4510-a50d-5caf35296164">

Reply:
I see.  For an empty file, we probably need to handle it differently.

Reply:
well! I am testing in our environment to see if there are still problems. 

Before that, I will convert this PR to draft

Reply:
Just taking a quick look. I think simply removing the `MemoizedSupplier` (just calling `MD5Hash#getDigester` every time) should technically work. From what I see, `MemoizedSupplier` only call the `MessageDigest#reset` at the first lazy instantiation. The subsequent `MemoizedSupplier` invocation will just return the `MessageDigest` without calling the `MessageDigest#reset`.

Ozone encountered some thread local MD5 issue (https://github.com/apache/ozone/pull/6435), might be related.

Reply:
> Just taking a quick look. I think simply removing the `MemoizedSupplier` (just calling `MD5Hash#getDigester` every time) should technically work. From what I see, `MemoizedSupplier` only call the `MessageDigest#reset` at the first invocation. The subsequent `MemoizedSupplier` invocation will just return the `MessageDigest` instance without calling the `MessageDigest#reset` which might cause the mismatch issue.
> 
> Ozone encountered some thread local MD5 issue ([apache/ozone#6435](https://github.com/apache/ozone/pull/6435)), might be related.

Thank you for your comment!

I also think that whether to remove MemoizedSupplier or not is irrelevant to the problem. My idea is to use a new digester for each file to solve this problem well.

I have pushed the code to this PR, you can take a look

Reply:
Yes, instantiating a new digester is should also fine, but there might be more memory overhead and short lived objects compared to the ThreadLocal implementation (which might also be fine if the operation is infrequent).

If the issue can be resolved by simply removing the `MemoizedSupplier`, I think we can consider this instead of instantiating a new digester every time.

Reply:
> Just taking a quick look. I think simply removing the `MemoizedSupplier` (just calling `MD5Hash#getDigester` every time) should technically work. From what I see, `MemoizedSupplier` only call the `MessageDigest#reset` at the first invocation. The subsequent `MemoizedSupplier` invocation will just return the `MessageDigest` instance without calling the `MessageDigest#reset` which might cause the mismatch issue.
> 
> Ozone encountered some thread local MD5 issue ([apache/ozone#6435](https://github.com/apache/ozone/pull/6435)), might be related.



> Just taking a quick look. I think simply removing the `MemoizedSupplier` (just calling `MD5Hash#getDigester` every time) should technically work. From what I see, `MemoizedSupplier` only call the `MessageDigest#reset` at the first invocation. The subsequent `MemoizedSupplier` invocation will just return the `MessageDigest` instance without calling the `MessageDigest#reset` which might cause the mismatch issue.

<img width="1140" alt="image" src="https://github.com/user-attachments/assets/e93e5d95-bfbb-4ade-9468-f5cb0a8f2a50">
In fact, digester.get ().reset () will be executed every time the first chunk is received.

In addition, I have done JMH testing, and it takes 0.5s for new 1000000 digests, but in fact, we do not send so many files so frequently, and this overhead is not a problem